applications/nrf5340_audio/src/audio_module/amod_api.c

```diff
@@ -0,0 +1,1136 @@
+				continue;
+			}
+
+			/* Send output audio object to next module(s) */
+			if (!sys_slist_is_empty(&hdl->hdl_dest_list)) {
+				k_mutex_lock(&hdl->dest_mutex, K_FOREVER);
+				k_sem_init(&hdl->sem, 0, hdl->dest_count);
+
+				SYS_SLIST_FOR_EACH_CONTAINER(&hdl->hdl_dest_list, hdl_to, node) {
+					ret = data_send(hdl, hdl_to, &out_msg->object,
+							&object_release_cb);
+					if (ret) {
+						LOG_DBG("Failed to send to module %s from %s",
+							hdl_to->name, hdl->name);
+					} else {
+						k_sem_give(&hdl->sem);
+					}
+				}
+
+				k_mutex_unlock(&hdl->dest_mutex);
+			} else {
+				/* Send output audio object back to module sending the data */
+				ret = data_fifo_block_lock(&hdl->out_msg, (void **)&out_msg,
+							   sizeof(struct _amod_in_message));
+				if (ret) {
+					clean_up(hdl, &in_msg, &out_msg, &data);
+
+					LOG_DBG("Module %s failed to return output buffer, ret %d",
+						hdl->name, ret);
+					continue;
+				}
+				LOG_DBG("Data sent from module %s", hdl->name);
+			}
+
+			if (in_msg->response_cb != NULL) {
+				in_msg->response_cb(in_msg->tx_handle, in_msg->object);
+			}
+
+		} else {
+			LOG_DBG("No process function for module %s, discarding input", hdl->name);
+		}
+
+		data_fifo_block_free(&hdl->in_msg, (void **)&in_msg);
+	}
+
+	return 0;
+}
+/**
+ * @brief  Function for querying the resources required for a module.
+ */
+int amod_query_resource(struct amod_parameters *parameters,
+			struct amod_configuration *configuration)
+{
+	struct _amod_functions *functions =
+		(struct _amod_functions *)parameters->description->functions;
+	int ret;
+	int size = 0;
+
+	ret = validate_parameters(parameters);
+	if (ret) {
+		LOG_DBG("Invalid parameters for module, returned %d", ret);
+		return ret;
+	}
+
+	if (functions->query_resource != NULL) {
+		size = functions->query_resource(configuration);
+		if (size < 0) {
+			LOG_DBG("Failed query resource for module %s, ret %d",
+				parameters->description->name, size);
+			return -EFAULT;
+		}
+	} else {
+		LOG_DBG("No query resource function for module %s", parameters->description->name);
+	}
+
+	size += WB_UP(sizeof(struct _amod_handle));
+
+	return size;
+};
+
+/**
+ * @brief  Function for opening a module.
+ */
+int amod_open(struct amod_parameters *parameters, struct amod_configuration *configuration,
+	      char *name, char *in_msg_block, char *out_msg_block, char *data_block,
+	      size_t data_size, uint32_t data_num, struct amod_handle *handle)
+{
+	struct _amod_handle *hdl = (struct _amod_handle *)handle;
+	int ret;
+
+	if (handle == NULL) {
+		LOG_DBG("Input handle parameter is NULL for module %s in the open function", name);
+		return -EINVAL;
+	}
+
+	if (hdl->state == AMOD_STATE_OPEN) {
+		LOG_DBG("Module %s is already open", hdl->name);
+		return -EALREADY;
+	}
+
+	ret = validate_parameters(parameters);
+	if (ret) {
+		LOG_DBG("Invalid parameters for module, ret %d", ret);
+		return ret;
+	}
+
+	memset(handle, 0, WB_UP(sizeof(struct _amod_handle)));
```

Comment:
May we have an issue here? Lets say that the size is 6 so `WB_UP` returns 8.

Reply:
When the data block is size is assigned for the handle in amod_query_resource() it is also rounded up and this is to ensure all memory is intialised to zero, even any extra bytes.